Assists: send user ID over the network rather than entindex

squeek502 opened the pull request:
Alternate solution for #318

squeek502 committed:
Assists: send user ID over the network rather than entindex

AfterShockFF commented on cl_dll/ff/ff_hud_deathnotice.cpp:
```diff
@@ -706,14 +706,12 @@ void CHudDeathNotice::FireGameEvent( IGameEvent * event )
-			//if ( !pFFPlayer )
-			Q_strcpy( deathMsg.Assister.szName, g_PR->GetPlayerName( killassisterIdx ) );
-			deathMsg.Assister.iEntIndex = killassisterIdx;
+			killAssister = engine->GetPlayerForUserID(killAssister);
```
I think this would be more explicit if we keep the concepts of userId and playerIndex separate and stored in different variables. It would help prevent future coders getting confused and making a bug.

killAssisterUserId and killAssisterIndex for example.

AfterShockFF commented on game_shared/multiplay_gamerules.cpp:
```diff
@@ -810,7 +810,7 @@ bool CMultiplayRules::IsMultiplayer( void )
 				if ( pTopAssister )
 				{
-					event->SetInt("killassister", pTopAssister->playerIndex );
+					event->SetInt("killassister", pTopAssister->GetUserID() );
```
Does this compile? I'm confused - RecentAttackInfo doesnt have a GetUserId function on?

squeek502 replied:
Haha I'm a dummy

squeek502 committed:
cleanup: Make userid/entindex distinction clearer

squeek502 committed:
Use CHandle to store recent attacker info rather than entindex

"An entity index is not reliable between frames, as it can be freed from one entity and and reallocated to another at any time. Use CHandle for long-term storage.

An entity index is a unique integer given to every entity by the engine. It allows the same entity to be referred to across library boundaries, but will be different on the client and server unless an edict is used. Within the client and server, pointers to CBaseEntity are more useful."

AfterShockFF commented on dlls/ff/ff_player.cpp:
```diff
@@ -7908,14 +7908,13 @@ void CFFPlayer::AddRecentAttacker( const CTakeDamageInfo &dmgInfo )
 	for ( int i = 0; i < m_recentAttackers.Count( ); ++i )
 	{
-		if ( m_recentAttackers[i].playerIndex == attackerIdx || pAttacker == m_recentAttackers[i].pFFPlayer )
+		if ( pAttacker == m_recentAttackers[i].hPlayer )
```
You don't have to do a .Get here then?

squeek502 replied:
CHandle has an == override that makes this work

AfterShockFF commented on dlls/ff/ff_player.h:
```diff
@@ -102,18 +102,15 @@ struct LocationInfo
+	RecentAttackerInfo(CFFPlayer *pPlayer, float dmg, float ts)
 	{
-		playerIndex = idx;
+		hPlayer = pPlayer;
```
Does this casting happen by magic then?

squeek502 replied:
CHandle has an assignment override that makes this work

DexterHaslem committed:
rename MAX_ASSIST_TIME_MS -> MAX_ASSIST_TIME_SECS
i forgot gpGlobals->curtime is in float second
this made being included in assists last way too long
